Translate guides/tutorial/part-2/ember-data.md (#242)
DazzlingFugu/ember-fr-guides-source: guides/tutorial/part-2/ember-data.md

> Dans ce chapitre, nous allons nous atteler à supprimer du code dupliqué dans nos gestionnaires de routes (_route handlers_), en utilisant Ember Data pour gérer nos données. Visuellement, le résultat final sera exactement comme avant&nbsp;:

Reviewer: Not necessarily requesting a change, eventually but not mandatory, but do we want to stay consistent when using "EmberData" or "Ember Data"?

The Ember.js team uses the format `EmberData`:
https://github.com/emberjs/data/

Author: Oh, that's very interesting. There is a PR in the English guides that [replaces all the references to "Ember Data" with "EmberData"](https://github.com/ember-learn/guides-source/pull/1896/files), and a while ago I did a "catchup English" that takes this into account. But the tutorial part was not modified in this PR because the tutorial part is generated from elsewhere on the English side.

}
```

<!-- spell ignore -->

> Cette duplication a occasionné de la "dette technique", rendant notre code plus difficile à maintenir sur le long terme. Par exemple, si nous voulions changer la manière dont la récupération des données fonctionne, alors nous devrions éditer les deux routes `index` et `rental`. Si nous éditions un seul endroit mais oubliions l'autre, nous pourrions nous retrouver avec des bugs très subtils dans l'application&nbsp;! _Eeerk_&nbsp;_!_

Reviewer: ```suggestion
Cette duplication a occasionné de la "dette technique", rendant notre code plus difficile à maintenir sur le long terme. Par exemple, si nous voulions changer la manière dont la récupération des données fonctionne, alors nous devrions éditer les deux routes `index` et `rental`. Si nous éditions un seul endroit mais oublions l'autre, nous pourrions nous retrouver avec des bugs très subtils dans l'application&nbsp;! _Eeerk_&nbsp;_!_
```

Author: `oubliions` is correct, it's imparfait.

      "category": "Domaine",
      "bedrooms": 15,
      "image": "https://upload.wikimedia.org/wikipedia/commons/c/cb/Crane_estate_(5).jpg",
      "description": "Ce manoir ancien et spacieux se trouve sur un domaine de plus de 100 acres de collines et de forêts de séquoias denses."
    }
  }
}
```

> On accède aux attributs d'une instance de `RentalModel` en utilisant une notation standard avec un point, comme `model.title` ou `model.location.lat`. En plus des attributs déclarés avec `@attr`, l'instance a toujours un attribut _id_ implicite, qui sert d'identifiant unique à l'objet modèles et auquel on peut accéder avec `model.id`.

Reviewer: ```suggestion
On accède aux attributs d'une instance de `RentalModel` en utilisant une notation standard avec un point, comme `model.title` ou `model.location.lat`. En plus des attributs déclarés avec `@attr`, l'instance a toujours un attribut _id_ implicite, qui sert d'identifiant unique à l'objet modèle et auquel on peut accéder avec `model.id`.
```

> Les classes `Model` de Ember Data ne sont pas différentes des autres classes avec lesquelles nous avons travaillé jusqu'ici, dans le sens où elles font un endroit idéal pour définir des comportements personnalisés. Par exemple, on a déplacé dans la classe `RentalModel` notre logique de `type` (qui était une source de duplication majeure et non nécessaire dans nos gestionnaires de route) à l'intérieur d'un _getter_ (accesseur). Une fois que tout fonctionnera ici, nous irons faire un peu de ménage.

Reviewer: I would prefer `la classe est l'endroit idéal pour ...` over `la classe fait un endroit idéal pour ...`.

```suggestion
Les classes `Model` de Ember Data ne sont pas différentes des autres classes avec lesquelles nous avons travaillé jusqu'ici, dans le sens où elles sont l'endroit idéal pour définir des comportements personnalisés. Par exemple, on a déplacé dans la classe `RentalModel` notre logique de `type` (qui était une source de duplication majeure et non nécessaire dans nos gestionnaires de route) à l'intérieur d'un _getter_ (accesseur). Une fois que tout fonctionnera ici, nous irons faire un peu de ménage.
```

---

Not directly related, I see

> _getter_ (accesseur)
> EN term (FR term)

while I also see

> FR term (EN term)

which one should be used?

Here, should it be instead `accesseur (_getter_)` like in ```Le décorateur (_decorator_) `@attr` ``` from above?

Author: I think it's ok this way. "décorateur" is quite intuitive in French, but "getter" is more common than "accesseur". We have already used "_getter_ (accesseur)" in another file.

Les attributs déclarés avec le décorateur `@attr` fonctionnent avec l'_auto-track_ mentionné [dans un chapitre précédent](../../part-1/reusable-components/). Ainsi, on peut référencer librement les attributs d'un modèle dans un _getter_ (`this.category`), et Ember saura quand invalider le résultat.

## Tester les modèles

> Jusque-là, nous n'avions pas d'endroit idéal pour écrire les tests de la logique `type` des locations. Maintenant que nous lui avons trouvé un foyer dans la classe modèle, il est également plus facile de tester le comportement. Nous allons ajouter un fichier de test à l'aide du générateur `model-test`&nbsp;:

Reviewer: ```suggestion
Jusque-là, nous n'avions pas d'endroit idéal pour écrire les tests de la logique `type` des locations. Maintenant que nous lui avons trouvé un foyer dans la classe modèle, il est également plus facile de tester son comportement. Nous allons ajouter un fichier de test à l'aide du générateur `model-test`&nbsp;:
```

});
```

> Ce test de modèle est un [test unitaire (_unit test_)](../../../testing/testing-models/). Contrairement aux tests que nous avons écrit jusqu'à maintenant, celui-ci "n'affiche" rien. Il ne fait qu'instancier le modèle `rental` et le tester directement, en manipulant ses attributs et en faisant des assertions sur leur valeur.

Reviewer: ```suggestion
Ce test de modèle est un [test unitaire (_unit test_)](../../../testing/testing-models/). Contrairement aux tests que nous avons écrits jusqu'à maintenant, celui-ci "n'affiche" rien. Il ne fait qu'instancier le modèle `rental` et le tester directement, en manipulant ses attributs et en faisant des assertions sur leur valeur.
```

Il est important de noter qu'Ember Data fournit un [service `store`](../../../services/) (magasin, réserve), aussi connu sous le nom de _Ember Data store_. Dans notre test, l'API `this.owner.lookup('service:store')` accède au _store_ d'Ember Data. Le _store_ fournit une méthode `createRecord` (créer un enregistrement) qui instancie un modèle pour nous.

Exécuter les tests dans le navigateur confirme que tout fonctionne comme prévu&nbsp;:

<img src="/images/tutorial/part-2/ember-data/pass-1@2x.png" alt="Tous les tests passent !" width="1024" height="1024">

## Charger un modèle dans une route

> Très bien, maintenant que notre modèle est prêt, il est temps de refactorer nos gestionnaires de route pour utiliser Ember Data et supprimer la duplication de code&nbsp;!

Reviewer: Always tricky this one, on l'utilise à toutes les sauces, what about this:

```suggestion
Très bien, maintenant que notre modèle est prêt, il est temps de _refactor_ nos gestionnaires de route pour utiliser Ember Data et supprimer le code dupliqué&nbsp;!
```

Author: I wouldn't change "refactorer" here, it's pretty common in French-speaking environments.

Voilà qui fait beaucoup de théorie, mais est-ce que ça va fonctionner dans notre app&nbsp;? Exécutons les tests pour le découvrir&nbsp;!

<img src="/images/tutorial/part-2/ember-data/fail-1@2x.png" alt="Quelques tests échouent !" width="1024" height="960">

<!-- spell ignore -->
Nooon, quelques tests échouent&nbsp;! Cela dit, c'est chouette d'être ainsi informé des problèmes potentiels (Vive les tests de non-régression &nbsp;!)

Si on regarde le message d'erreur, il apparaît que le _store_ ne visite pas la bonne URL quand on récupère les données du serveur, conduisant à des erreurs 404. Plus précisément&nbsp;:

- Quand la requête `findAll('rental')` s'exécute, elle cherche à récupérer les données de `/rentals`, au lieu de `/api/rentals.json`.
- Quand la requête `find('rental', 'grand-old-mansion')` s'exécute, elle cherche à récupérer les données de `/rentals/grand-old-mansion`, au lieu de `/api/rentals/grand-old-mansion.json`.

Ok, très bien, donc il nous faut indiquer à Ember Data de requêter les données à la bonne adresse. Mais d'abord, comment Ember Data sait-il comment récupérer les données de notre serveur&nbsp;?

## Travailler avec des adaptateurs et des sérialiseurs

> Ember Data utilise une architecture [adapteur (_adapter_)](../../../models/customizing-adapters/) et [sérialiseur (_serializer_)](../../../models/customizing-serializers/). L'adapteur s'occupe de "comment" et "où" Ember Data doit récupérer les données du serveur, par exemple faut-il utiliser HTTP, HTTPS, des _WebSockets_ ou le _local storage_&nbsp;? quelles sont les URLs, les entêtes et les paramètres pour ces requêtes&nbsp;? Le sérialiseur, lui, est en charge de convertir les données retournées par le serveur en un format qu'Ember Data comprend.

Reviewer: ```suggestion
Ember Data utilise une architecture [adaptateur (_adapter_)](../../../models/customizing-adapters/) et [sérialiseur (_serializer_)](../../../models/customizing-serializers/). L'adaptateur s'occupe de "comment" et "où" Ember Data doit récupérer les données du serveur, par exemple faut-il utiliser HTTP, HTTPS, des _WebSockets_ ou le _local storage_&nbsp;? Quelles sont les URLs, les en-têtes et les paramètres pour ces requêtes&nbsp;? Le sérialiseur, lui, est chargé de convertir les données retournées par le serveur en un format qu'Ember Data comprend.
```

> L'idée est la suivante&nbsp;: en supposant que votre _backend_ expose un protocole cohérent et un format d'échange pour accéder à ses données, on peut alors écrire une seule paire d'adapteur-sérialiseur pour traiter toutes les données récupérées pour l'application entière.

Reviewer: I see a clear difference between

> traiter les données récupérées

and

> traiter les récupérations de données

```suggestion
L'idée est la suivante&nbsp;: en supposant que votre _backend_ expose un protocole cohérent et un format d'échange pour accéder à ses données, on peut alors créer une seule paire d'adaptateur-sérialiseur pour gérer toutes les récupérations de données pour l'application entière.
```

> Il se trouve que le format JSON:API est le protocole de données et format d'échange par défaut de Ember Data. Autrement dit, Ember Data fournit clé en main un adapteur et un sérialiseur JSON:API. C'est une très bonne nouvelle pour nous, puisque c'est aussi ce que notre serveur implémente. Quelle heureuse coïncidence&nbsp;!

Reviewer: ```suggestion
Il se trouve que le format JSON:API est le protocole de données et format d'échange par défaut de Ember Data. Autrement dit, Ember Data fournit clé en main un adaptateur et un sérialiseur JSON:API. C'est une très bonne nouvelle pour nous, puisque c'est aussi ce que notre serveur implémente. Quelle heureuse coïncidence&nbsp;!
```

> Cependant, comme mentionné ci-dessus, il y a quelques différences mineures entre le fonctionnement de notre serveur et les présomptions d'Ember Data. Nous pouvons personnaliser le comportement par défaut en définissant notre propre adapteur et sérialiseur&nbsp;:

Reviewer: ```suggestion
Cependant, comme mentionné ci-dessus, il y a quelques différences mineures entre le fonctionnement de notre serveur et les présomptions d'Ember Data. Nous pouvons personnaliser le comportement par défaut en définissant notre propre adaptateur et sérialiseur&nbsp;:
```

export default class ApplicationSerializer extends JSONAPISerializer {}
```

> Par convention, les adapteurs se trouvent dans `app/adapters`. De plus, l'adapteur nommé `application` est appelé "adapteur de l'application" (_application adapter_), et récupérera les données de tous les modèles de l'app.

Reviewer: ```suggestion
Par convention, les adaptateurs se trouvent dans `app/adapters`. De plus, l'adaptateur nommé `application` est appelé "adaptateur de l'application" (_application adapter_), et récupérera les données de tous les modèles de l'app.
```

De la même manière, les sérialiseurs se situent dans `app/serializers`. L'adapteur et le sérialiseur sont toujours ajoutés par paire. Nous avons ajouté un adapteur `application`, donc nous avons aussi ajouté le sérialiseur correspondant. Puisque les données JSON retournées par notre serveur sont conformes à JSON:API, le sérialiseur par défaut [`JSONAPISerializer`](https://api.emberjs.com/ember-data/release/classes/JSONAPISerializer) fonctionne très bien pour nous, pas besoin de le personnaliser.

Avec notre adapteur et notre sérialiseur en place, nos tests devraient passer à nouveau.

<img src="/images/tutorial/part-2/ember-data/pass-2@2x.png" alt="Une fois de plus, tous nos tests passent !" width="1024" height="1024">

> L'UI aussi fonctionne exactement comme avant, seulement avec moins de code&nbsp;!

Author: I will replace it with "interface utilisateur" without the English term because it's just one occurrence. In general, maybe we can go with _UI_ if the term is used a lot in the same file.
